contracts/src/token/erc20/extensions/erc4626.rs

```diff
@@ -0,0 +1,528 @@
+        //shares.mul_div(x, y, dominator, rounding)
+        U256::ZERO
+    }
+
+    fn _deposit(
+        &mut self,
+        caller: Address,
+        receiver: Address,
+        assets: U256,
+        shares: U256,
+    ) -> Result<(), Error> {
+        // If _asset is ERC-777, `transferFrom` can trigger a reentrancy BEFORE
+        // the transfer happens through the `tokensToSend` hook. On the
+        // other hand, the `tokenReceived` hook, that is triggered after the
+        // transfer, calls the vault, which is assumed not malicious.
+        //
+        // Conclusion: we need to do the transfer before we mint so that any
+        // reentrancy would happen before the assets are transferred and
+        // before the shares are minted, which is a valid state.
+        // slither-disable-next-line reentrancy-no-eth
+
+        self._asset._mint(receiver, shares)?;
+        evm::log(Deposit { sender: caller, owner: receiver, assets, shares });
+        Ok(())
+    }
+
+    fn _withdraw(
+        &mut self,
+        caller: Address,
+        receiver: Address,
+        owner: Address,
+        assets: U256,
+        shares: U256,
+    ) -> Result<(), Error> {
+        if caller != owner {
+            self._asset._spend_allowance(owner, caller, shares)?;
+        }
+
+        // If _asset is ERC-777, `transfer` can trigger a reentrancy AFTER the
+        // transfer happens through the `tokensReceived` hook. On the
+        // other hand, the `tokensToSend` hook, that is triggered before the
+        // transfer, calls the vault, which is assumed not malicious.
+        //
+        // Conclusion: we need to do the transfer after the burn so that any
+        // reentrancy would happen after the shares are burned and after
+        // the assets are transferred, which is a valid state.
+        self._asset._burn(owner, shares)?;
+        self._safe_erc20.safe_transfer(
+            contract::address(),
+            receiver,
+            assets,
+        )?;
+
+        evm::log(Withdraw { sender: caller, receiver, owner, assets, shares });
+        Ok(())
+    }
+
+    /// Offset of the decimals of the ERC-20 asset from the decimals of the
+    /// Vault.
+    ///
+    /// This value is used to calculate the number of shares that can be minted
+    /// for a given amount of assets, and to calculate the number of assets
+    /// that can be withdrawn for a given amount of shares.
+    ///
+    /// The value is set to 0 by default, which means that the decimals of the
+    /// ERC-20 asset and the Vault are the same.
+    ///
+    /// To change this value, you must override this function in your contract.
+    fn _decimals_offset(&self) -> u32 {
+        0
+    }
+}
+
+#[cfg(all(test, feature = "std"))]
+mod tests {
+
+    use super::Erc4626;
+
+    #[motsu::test]
+    fn can_get_max_mint(contract: Erc4626) {
+        let sender = msg::sender();
+        let max_mint = Erc4626::max_mint(&self, sender);
```

Comment:
```suggestion
        let max_mint = contract.max_mint(sender);
```
1. No need for an associated method call.
2. There is no `&self` that you can pass to the function call.
3. This does not compile.

Similar suggestions for [the other test](https://github.com/OpenZeppelin/rust-contracts-stylus/pull/465/files#diff-1a44e32d6bb9877830360c6f8f98175e7d4c01b6e50b30fed49c5a434e3e0236R523).

Please make sure the tests are compiling and running successfully before opening a PR.